internal/command/workdir/statestore_config_state.go

```diff
@@ -0,0 +1,180 @@
+
+package workdir
+
+import (
+	"encoding"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	version "github.com/hashicorp/go-version"
+	tfaddr "github.com/hashicorp/terraform-registry-address"
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/configs/configschema"
+	"github.com/hashicorp/terraform/internal/plans"
+	"github.com/zclconf/go-cty/cty"
+	ctyjson "github.com/zclconf/go-cty/cty/json"
+)
+
+var _ ConfigState[StateStoreConfigState] = &StateStoreConfigState{}
+
+// StateStoreConfigState describes the physical storage format for the state store
+type StateStoreConfigState struct {
+	Type      string          `json:"type"`     // State store type name
+	Provider  *Provider       `json:"provider"` // Details about the state-storage provider
+	ConfigRaw json.RawMessage `json:"config"`   // state_store block raw config, barring provider details
+	Hash      uint64          `json:"hash"`     // Hash of portion of configuration from config files
+}
+
+// Provider is used in the StateStoreConfigState struct to describe the provider that's used for pluggable
+// state storage. The data inside should mirror an entry in the dependency lock file.
+// This is NOT state of a `provider` configuration block, or an entry in `required_providers`.
+type Provider struct {
+	Version *version.Version `json:"version"` // The specific provider version used for the state store. Should be set using a getproviders.Version, etc.
+	Source  *Source          `json:"source"`  // The FQN/fully-qualified name of the provider.
+}
+
+type Source struct {
+	Provider tfaddr.Provider `json:"-"` // For un/marshaling purposes this needs to be exported, but it's omitted from JSON
+}
+
+// TODO - move this upstream
+var _ encoding.TextMarshaler = &Source{}
+var _ encoding.TextUnmarshaler = &Source{}
+
+// MarshalText defers to the internal terraform-registry-address Provider's String method
+func (p *Source) MarshalText() ([]byte, error) {
+	fqn := p.Provider.String()
+	return []byte(fqn), nil
+}
+
+// UnmarshalText uses existing methods for parsing `source` values from `required_providers` config to
+// convert a source string into a terraform-registry-address Provider
+func (p *Source) UnmarshalText(text []byte) error {
+	input := string(text)
+	provider, diag := addrs.ParseProviderSourceString(input)
+	if diag != nil {
+		return fmt.Errorf("error unmarshaling provider source from backend state file: %s", diag.ErrWithWarnings())
+	}
+	p.Provider = provider
+	return nil
+}
+
+// Empty returns true if there is no active state store.
+func (s *StateStoreConfigState) Empty() bool {
+	return s == nil || s.Type == ""
+}
+
+// Validate returns true if there are no missing expected values, and
+// important values have been validated, e.g. FQNs. When the config is
+// invalid an error will be returned.
+func (s *StateStoreConfigState) Validate() error {
+
+	// TODO - refactor once upstreaming of parsing logic complete
```

Comment:
Code changes required once updates to `hashicorp/terraform-registry-address` happen

Reply:
https://github.com/hashicorp/terraform/pull/37230